music-player/src/reducers.tsx

```diff
@@ -0,0 +1,103 @@
+import {
+  ADD_PLAYLIST_ACTION_TYPE,
+  ADD_SONG_TO_PLAYLIST_ACTION_TYPE,
+  CHANGE_PLAYLIST_ACTION_TYPE,
+  PLAY_NEXT_SONG_ACTION_TYPE,
+  PLAY_PREV_SONG_ACTION_TYPE,
+  REMOVE_SONG_FROM_PLAYLIST_ACTION_TYPE,
+} from './actions';
+import { songsNamesList } from './data';
+import { Action, SongsListState, State } from './types';
+import { routerReducer } from 'react-router-redux';
+import { combineReducers } from 'redux';
+import { viewport } from './forUrl/viewport';
+
+const initialSongsListState: SongsListState = {
+  currentSongIndex: 0,
+  currentPlaylistName: 'All',
+  playlists: {
+    All: songsNamesList.map((name, index) => index),
+    FirstSongs: [0, 1],
+  },
+  isValidURL: true
+};
+
+const songsListStateReducer = (songsListState: SongsListState = initialSongsListState, action: Action) => {
```

Comment:
Also, referring to my other comment, the 'Action' type here doen't provide any real guarantees against using undefined values.